File: sdk/nodejs/src/resources/cloud_instance.js
"use strict";

const { BaseResource } = require("./base");

class CloudInstanceResource extends BaseResource {
  async list(serviceId) {
    return this._http.get("/api/service/{id}/vms");
  }

  async create(serviceId, data) {
    return this._http.post("/api/service/{id}/vms", data);
  }

  async get(serviceId, vmId) {
    return this._http.get("/api/service/{id}/vms/{vmid}");
  }

  async delete(serviceId, vmId) {
    return this._http.delete("/api/service/{id}/vms/{vmid}");
  }

  async update(serviceId, vmId, data) {
    return this._http.put("/api/service/{id}/vms/{vmid}", data);
  }

  async stop(serviceId, vmId) {
    return this._http.post("/api/service/{id}/vms/{vmid}/stop");
  }

  async start(serviceId, vmId) {
    return this._http.post("/api/service/{id}/vms/{vmid}/start");
  }

  async reboot(serviceId, vmId) {
    return this._http.post("/api/service/{id}/vms/{vmid}/reboot");
  }

  async resetPassword(serviceId, vmId, data) {
    return this._http.post("/api/service/{id}/vms/{vmid}/resetpwd", data);
  }

  async rebuild(serviceId, vmId, data) {
    return this._http.post("/api/service/{id}/vms/{vmid}/rebuild", data);
  }

  async listIps(serviceId, vmId) {
    return this._http.get("/api/service/{id}/vms/{vmid}/ips");
  }

  async addIp(serviceId, vmId, data) {
    return this._http.post("/api/service/{id}/vms/{vmid}/ips", data);
  }

  async listInterfaces(serviceId, vmId) {
    return this._http.get("/api/service/{id}/vms/{vmid}/interfaces");
  }

  async getCpuUsage(serviceId, vmId) {
    return this._http.get("/api/service/{id}/vms/{vmid}/usage/cpu");
  }

  async getNetUsage(serviceId, vmId) {
    return this._http.get("/api/service/{id}/vms/{vmid}/usage/net");
  }

  async getDiskUsage(serviceId, vmId) {
    return this._http.get("/api/service/{id}/vms/{vmid}/usage/disk");
  }

  async listStorage(serviceId, vmId) {
    return this._http.get("/api/service/{id}/vms/{vmid}/storage");
  }

  async resizeStorage(serviceId, vmId, diskId, data) {
    return this._http.put("/api/service/{id}/vms/{vmid}/storage/{diskid}", data);
  }

  async listImages(serviceId) {
    return this._http.get("/api/service/{id}/images");
  }

  async createImage(serviceId, data) {
    return this._http.post("/api/service/{id}/images", data);
  }

  async getInterfaceIps(serviceId, vmId, iface) {
    return this._http.get(`/api/service/{id}/vms/{vmid}/interfaces/${iface}/ips`);
  }

  async deleteIp(serviceId, vmId, ipId) {
    return this._http.delete("/api/service/{id}/vms/{vmid}/ips/{ipid}");
  }

  async getRdns(serviceId, vmId) {
    return this._http.get("/api/service/{id}/vms/{vmid}/rdns");
  }

  async setRdns(serviceId, vmId, data) {
    return this._http.post("/api/service/{id}/vms/{vmid}/rdns", data);
  }

  async rebuildNetwork(serviceId, vmId) {
    return this._http.post("/api/service/{id}/vms/{vmid}/rebuild_network");
  }

}

module.exports = { CloudInstanceResource };

File: sdk/csharp/Resources/CloudInstance.cs
namespace HiTechCloud.SDK.Resources;

public class CloudInstance : BaseResource
{
    public CloudInstance(HttpClientWrapper http) : base(http) { }

    public async Task<Dictionary<string, object?>> VmsAsync(int id)
    {
        return await Http.GetAsync(string.Format("/api/service/{id}/vms", id));
    }

    public async Task<Dictionary<string, object?>> CreateVmAsync(int id, Dictionary<string, object> data)
    {
        return await Http.PostAsync(string.Format("/api/service/{id}/vms", id), data);
    }

    public async Task<Dictionary<string, object?>> VmAsync(int id, int vmid)
    {
        return await Http.GetAsync(string.Format("/api/service/{id}/vms/{vmid}", id, vmid));
    }

    public async Task<Dictionary<string, object?>> DeleteVmAsync(int id, int vmid)
    {
        return await Http.DeleteAsync(string.Format("/api/service/{id}/vms/{vmid}", id, vmid));
    }

    public async Task<Dictionary<string, object?>> UpdateVmAsync(int id, int vmid, Dictionary<string, object> data)
    {
        return await Http.PutAsync(string.Format("/api/service/{id}/vms/{vmid}", id, vmid), data);
    }

    public async Task<Dictionary<string, object?>> StopVmAsync(int id, int vmid)
    {
        return await Http.PostAsync(string.Format("/api/service/{id}/vms/{vmid}/stop", id, vmid), null);
    }

    public async Task<Dictionary<string, object?>> StartVmAsync(int id, int vmid)
    {
        return await Http.PostAsync(string.Format("/api/service/{id}/vms/{vmid}/start", id, vmid), null);
    }

    public async Task<Dictionary<string, object?>> RebootVmAsync(int id, int vmid)
    {
        return await Http.PostAsync(string.Format("/api/service/{id}/vms/{vmid}/reboot", id, vmid), null);
    }

    public async Task<Dictionary<string, object?>> ResetPasswordAsync(int id, int vmid, Dictionary<string, object> data)
    {
        return await Http.PostAsync(string.Format("/api/service/{id}/vms/{vmid}/resetpwd", id, vmid), data);
    }

    public async Task<Dictionary<string, object?>> RebuildVmAsync(int id, int vmid, Dictionary<string, object> data)
    {
        return await Http.PostAsync(string.Format("/api/service/{id}/vms/{vmid}/rebuild", id, vmid), data);
    }

    public async Task<Dictionary<string, object?>> VmIpsAsync(int id, int vmid)
    {
        return await Http.GetAsync(string.Format("/api/service/{id}/vms/{vmid}/ips", id, vmid));
    }

    public async Task<Dictionary<string, object?>> AddVmIpAsync(int id, int vmid, Dictionary<string, object> data)
    {
        return await Http.PostAsync(string.Format("/api/service/{id}/vms/{vmid}/ips", id, vmid), data);
    }

    public async Task<Dictionary<string, object?>> VmInterfacesAsync(int id, int vmid)
    {
        return await Http.GetAsync(string.Format("/api/service/{id}/vms/{vmid}/interfaces", id, vmid));
    }

    public async Task<Dictionary<string, object?>> CpuUsageAsync(int id, int vmid)
    {
        return await Http.GetAsync(string.Format("/api/service/{id}/vms/{vmid}/usage/cpu", id, vmid));
    }

    public async Task<Dictionary<string, object?>> NetUsageAsync(int id, int vmid)
    {
        return await Http.GetAsync(string.Format("/api/service/{id}/vms/{vmid}/usage/net", id, vmid));
    }

    public async Task<Dictionary<string, object?>> DiskUsageAsync(int id, int vmid)
    {
        return await Http.GetAsync(string.Format("/api/service/{id}/vms/{vmid}/usage/disk", id, vmid));
    }

    public async Task<Dictionary<string, object?>> StorageAsync(int id, int vmid)
    {
        return await Http.GetAsync(string.Format("/api/service/{id}/vms/{vmid}/storage", id, vmid));
    }

    public async Task<Dictionary<string, object?>> UpdateStorageAsync(int id, int vmid, int diskid, Dictionary<string, object> data)
    {
        return await Http.PutAsync(string.Format("/api/service/{id}/vms/{vmid}/storage/{diskid}", id, vmid, diskid), data);
    }

    public async Task<Dictionary<string, object?>> ImagesAsync(int id)
    {
        return await Http.GetAsync(string.Format("/api/service/{id}/images", id));
    }

    public async Task<Dictionary<string, object?>> CreateImageAsync(int id, Dictionary<string, object> data)
    {
        return await Http.PostAsync(string.Format("/api/service/{id}/images", id), data);
    }

    public async Task<Dictionary<string, object?>> InterfaceIpsAsync(int id, int vmid, string iface)
    {
        return await Http.GetAsync(string.Format("/api/service/{id}/vms/{vmid}/interfaces/{iface}/ips", id, vmid, iface));
    }

    public async Task<Dictionary<string, object?>> DeleteVmIpAsync(int id, int vmid, int ipid)
    {
        return await Http.DeleteAsync(string.Format("/api/service/{id}/vms/{vmid}/ips/{ipid}", id, vmid, ipid));
    }

    public async Task<Dictionary<string, object?>> VmRdnsAsync(int id, int vmid)
    {
        return await Http.GetAsync(string.Format("/api/service/{id}/vms/{vmid}/rdns", id, vmid));
    }

    public async Task<Dictionary<string, object?>> SetVmRdnsAsync(int id, int vmid, Dictionary<string, object> data)
    {
        return await Http.PostAsync(string.Format("/api/service/{id}/vms/{vmid}/rdns", id, vmid), data);
    }

    public async Task<Dictionary<string, object?>> RebuildNetworkAsync(int id, int vmid)
    {
        return await Http.PostAsync(string.Format("/api/service/{id}/vms/{vmid}/rebuild_network", id, vmid), null);
    }
}
